### ISSUE TITLE
Segmentation fault with small repro

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#22750]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: archlinux
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:n/a
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:b'v1.9.0-rc2-5276-ge57874169f' 1.12.0-dev20181004
- **Python version**:3.6
- **Bazel version (if compiling from source)**:n/a
- **GCC/Compiler version (if compiling from source)**:n/a
- **CUDA/cuDNN version**:9.0
- **GPU model and memory**:1080Ti
- **Exact command to reproduce**:below

This code:
```python
import tensorflow as tf
import numpy as np

def f(boxes, scores):
    def f(X):
        prob, box = X
        output_shape = tf.shape(prob)
        ids = tf.reshape(tf.where(prob > 0.05), [-1])
        prob = tf.gather(prob, ids)
        box = tf.gather(box, ids)
        # prob = tf.Print(prob, [box, prob], summarize=100, message='boxandprob')
        selection = tf.image.non_max_suppression(box, prob, 100, 0.5)
        selection = tf.to_int32(tf.gather(ids, selection))
        selection = tf.Print(selection, [ids, selection], summarize=100, message='ids_selection_2')
        sorted_selection = -tf.nn.top_k(-selection, k=tf.size(selection))[0]
        mask = tf.sparse_to_dense(
            sparse_indices=sorted_selection,
            output_shape=output_shape,
            sparse_values=True,
            default_value=False)
        return mask

    masks = tf.map_fn(f, (scores, boxes), dtype=tf.bool, parallel_iterations=10)     # #cat x N
    return masks

with tf.device('/gpu:0'):
    boxes = tf.placeholder(tf.float32, (80, None, 4), name='boxes')
    scores = tf.placeholder(tf.float32, (80, None), name='scores')
    outs = f(boxes, scores)

config = tf.ConfigProto()
config.allow_soft_placement = True
sess = tf.Session(config=config)
data = dict(np.load('debug.npz'))
for k in range(1000):
    sess.run(outs, feed_dict={boxes: data['boxes'].transpose(1, 0, 2)[1:, :, :], scores: data['scores'][:, 1:].T})
    print(k)
```
causes segmentation fault on tf-nightly-gpu, as well as tensorflow-gpu==1.11.0. It works on 1.10.
It needs the data file `debug.npz` here: 
[debug.zip](https://github.com/tensorflow/tensorflow/files/2448247/debug.zip)

Note:
1. I tested on two machines, an error happen in >90% runs.
2. The code was distilled from the bug report [here](https://github.com/tensorpack/tensorpack/issues/919). The original bug report does not always segfault, but occasionally crash with other __different__ unreasonable TF internal errors, such as:
```
InvalidArgumentError (see above for traceback): scores has incompatible shape
         [[node map/while/non_max_suppression/NonMaxSuppressionV3 (defined at bug.py:15)  = NonMaxSuppressionV3[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](map/while/Gather
V2_1/_29, map/while/GatherV2/_31, map/while/non_max_suppression/NonMaxSuppressionV3/max_output_size/_33, map/while/non_max_suppression/iou_threshold/_35, map/while/non_max_suppression/score_thresh
old/_37)]]
``` 
```
2018-10-04 14:59:14.736180: F tensorflow/core/common_runtime/bfc_allocator.cc:458] Check failed: c->in_use() && (c->bin_num == kInvalidBinNum)                                                     
```
```
2018-10-04 14:59:49.523436: F tensorflow/core/common_runtime/bfc_allocator.cc:380] Check failed: h != kInvalidChunkHandle 
```
```
2018-10-05 00:12:03.720295: F ./tensorflow/core/framework/tensor.h:643] Check failed: new_num_elements == NumElements() (39 vs. 0)
```
```

InvalidArgumentError (see above for traceback): indices[1] = [0] is repeated
         [[{{node map/while/SparseToDense}} = SparseToDense[T=DT_BOOL, Tindices=DT_INT32, _class=["loc:@map/while/TensorArrayWrite/TensorArrayWriteV3"], validate_indices=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](map/while/Neg_1/_51, map/while/Shape/_53, map/while/SparseToDense/sparse_values/_55, map/while/SparseToDense/default_value/_57)]]
         [[{{node map/while/SparseToDense/sparse_values/_54}} = _Send[T=DT_BOOL, client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_111_map/while/SparseToDense/sparse_values", _device="/job:localhost/replica:0/task:0/device:GPU:0"](map/while/SparseToDense/sparse_values)]]
```
After distilled to this small repro, it seems to mostly do segfault. But the above error messages might help. Seems like a memory corruption.

**Comment by ppwwyyxx** [comment]
@tayo 
I believe your commit https://github.com/tensorflow/tensorflow/commit/8566d9e6fa7dbe3660339befe8b0a3344d24ef2b#diff-6731fe0e9dae6d68dca55b2d50d32c06R320 about NMS op causes this bug.

The input `Tensor`s of a `OpKernel::Compute` should not be stored as members of the `OpKernel`. This effectively make `OpKernel::Compute` not thread-safe and crash with my sample code above.

**Comment by tayo** [comment]
Hi Yuxin,

I was just made aware of this issue recently.  I am working on a fix and
will push this out soon.  Thanks for bringing this to attention!

-Tayo

**Comment by harshini-gadige** [comment]
@tayo Hi, any update on the fix ? Could you please share the PR# here.

**Comment by tayo** [comment]
Fix is under review.

**Comment by ppwwyyxx** [comment]
@tayo I think the fix is not correct and does not fix the problem. (It may avoid the crash, but may end up giving wrong results).
The bug still exists: `Compute()` should be thread-safe. It may be called concurrently by many callers, as quoted from tensorflow's source code:
https://github.com/tensorflow/tensorflow/blob/4a51abb1f935818b6a130f71b340b29379e52a07/tensorflow/core/framework/op_kernel.h#L90-L93

When you store a pointer to a Tensor in `OpKernel::Compute`, the `Compute` method become not thread-safe. If called concurrently by many threads, thread A set pointer to input A, then thread B might overwrite the pointer to input B, then thread A will access the wrong input B.

**Comment by ppwwyyxx** [comment]
It does not avoid the crash either. After compiling TF with latest master from github, the same code above still segfault.

**Comment by tayo** [comment]
Thanks. Issue reopened to look more closely.

**Comment by tayo** [comment]
@ppwwyyxx Thanks for pointing this out.

**Comment by ppwwyyxx** [comment]
Thanks for the fix!

**Comment by boyuangong** [comment]
Hi, I still run into this problem with the most recent version.
`import time
from model import *
from utils import *
import DMaskRCNN as model
import coco
from extract import parse_annotation, resize_bbox
import skimage.io

# Root directory of the project
root_dir = os.getcwd()

# Directory to save logs and trained model
model_dir = os.path.join(root_dir, "domain_logs")

# Local path to trained weights file
model_path = os.path.join(root_dir, "mask_rcnn_domain.h5")

test_image_dir = os.path.join(root_dir, "RTTS/JPEGImages")
test_9411_dir = os.path.join(root_dir, "RTTS/9411result")
test_result_dir = os.path.join(root_dir, "RTTS/result")
test_aod_dir = os.path.join(root_dir, "RTTS/AOD_Dehaze")
# test_MSCC_dir = os.path.join(root_dir, "RTTS/MSCCdehaze")
test_file_dir = os.path.join(root_dir, "RTTS/Annotations")
test_image_names = next(os.walk(test_image_dir))[2]
test_image_names.sort()

COCO_DIR = "COCO"
config = coco.CocoConfig()


# Override the training configurations with a few
# changes for inferencing.
class InferenceConfig(config.__class__):
    # Run detection on one image at a time
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1


config = InferenceConfig()
# config.display()

# Get the first available GPU
DEVICE_ID_LIST = GPUtil.getFirstAvailable()
DEVICE_ID = DEVICE_ID_LIST[0]  # grab first element from list

# Set CUDA_VISIBLE_DEVICES to mask out all other GPUs than the first available device id
os.environ["CUDA_VISIBLE_DEVICES"] = str(DEVICE_ID)


# Training dataset. Use the training set and 35K from the
# validation set, as as in the Mask RCNN paper.


def mAP_test(image_dir, file_dir):
    APs = []
    t1 = time.time()
    image_names = next(os.walk(image_dir))[2]
    image_names.sort()
    print(len(image_names))
    i = 0
    for image_name in image_names:
        # print(image_name)
        try:
            image = skimage.io.imread(os.path.join(image_dir, image_name))
            image, window, scale, padding = utils.resize_image(
                image,
                min_dim=config.IMAGE_MIN_DIM,
                max_dim=config.IMAGE_MAX_DIM,
                padding=config.IMAGE_PADDING)
            if image_name.find('AOD') != -1:
                image_name = image_name.replace('_AOD-Net', '')
            if image_name.find('dehaze') != -1:
                image_name = image_name.replace('_dehazed', '')
            gt_class_id, ori_gt_bbox = parse_annotation(image_name, file_dir)
            gt_bbox = resize_bbox(ori_gt_bbox, scale, padding)
            results = domain_model.detect([image], verbose=0)
            r = results[0]
            AP, precisions, recalls, overlaps = \
                utils.compute_box_ap(gt_bbox, gt_class_id,
                                     r["rois"], r["class_ids"], r["scores"])
            APs.append(AP)
        except:
            continue
        if i % 100 == 0:
            print("current i:", i)
        i = i+1
        if i > 4322:
            print("overflow")
            break

    t2 = time.time() - t1
    print("using time: ", t2)
    print("mAP: ", np.mean(APs))
    return np.mean(APs)


print('Build the mask rcnn model')
domain_model = model.Domain_MaskRCNN(mode="inference", model_dir=model_dir,
                                     config=config)
print('loading weights from: ', model_path)
domain_model.load_weights(model_path, by_name=True)

print('calculating the mAP from dehazed MSCC dataset')
cur_mAP2 = mAP_test(test_result_dir, test_file_dir)
print('current mAP is:', cur_mAP2)

del domain_model
K.clear_session()
`

**Comment by jendelel** [comment]
I ran into the same issue with TF 1.12. It is not deterministic and fails in about 60% of cases.

This is my error message:
```
F tensorflow/core/common_runtime/bfc_allocator.cc:458] Check failed: c->in_use() && (c->bin_num == kInvalidBinNum) 
```

**Comment by ppwwyyxx** [comment]
The bug has not been fixed in TF 1.12 and will be fixed in later version.

**Comment by tayo** [comment]
> The bug has not been fixed in TF 1.12 and has been fixed in later versions.

Ah yes, apparently this was not pulled into 1.12.  If this issue is blocking anyone, one suggestion is to try TF nightly, pending the next TF release.

**Comment by ppwwyyxx** [comment]
It is fixed in the latest release (1.13rc0).

**Comment by sharathpuranik** [comment]
> It is fixed in the latest release (1.13rc0).

Thanks a lot. Will upgrade to 1.13rc0 and test it out. 

`2018-10-04 14:59:14.736180: F tensorflow/core/common_runtime/bfc_allocator.cc:458] Check failed: c->in_use() && (c->bin_num == kInvalidBinNum)
`
I am getting the same exception in tensorflow serving and was considering compiling the nightly myself.

**Comment by learnermaxRL** [comment]
Is it working,I get this error when running faster rcnn in Multiprocessing environment:

Check failed: c->in_use() && (c->bin_num == kInvalidBinNum)

**Comment by geek101** [comment]
I noticed that this issue does not occur with tf-nightly-gpu `1.13.0.dev20190208` when compared to `1.12`.

**Comment by wronk** [comment]
Also can confirm that pulling the 1.13.0 tf-nightly-gpu build solved this issue for me

**Comment by codexponent** [comment]
This error had troubled me for days. I also tried with the nightly tf build, it worked but it cost me hard cause I had to update so many dependencies as well as the CUDA version (which I shouldn't have). So, I reverted back and tried to solve the problem. 
Actually the problem was that i was running the training command from the script. Weird. This had not been the problem for me before.
Example: running python training_something.py from my_script.sh. So running python training_something.py **directly** solved my problem(completly discarding my_script.sh). Hope it helped you guys too.
